### ISSUE TITLE
Bitbucket `is_first_time` can report incorrectly

### ISSUE THREAD

**Issue by mcwarman** [orhun/git-cliff#906]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the bug

The Bitbucket first time does not report correctly when a user has made a commit in another as well as there first merged contribution.

We had a scenario where a first time contributor wasn't being reported in the change log. Looking into it the Commits API from bitbucket it returns all commits across all branches. 

> `GET /repositories/{workspace}/{repo_slug}/commits/`
>
> Returns all commits in the repo in topological order (newest commit first). All branches and tags are included (similar to git log --all).

The user has also made another commit in another branch which was there first commit in the repo, but not there first commit on the branch being released.

Reference: <https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-commits-get>

```bash
curl https://api.bitbucket.org/2.0/repositories/<org>/<repo>/commits?include=<branch>
```

### Steps To Reproduce

1. New user commits and pushes branch
2. New user commits and pushes branch for second change
3. Merge change from item 2
4. `git-cliff --current --context --bitbucket-repo "org/repo"`

### Expected behavior

The `is_first_time` value is updated based on users contributions to the current branch (not the repo).

### Software information

- Operating system: WSL - Debian GNU/Linux 12 (bookworm)
- Rust version: N/A
- Project version: 2.6.1

### Additional context

I guess the fix is to make the current branch context available in the `Remote` struct so it can be used in the API call.

https://github.com/orhun/git-cliff/blob/508a97edb088f77d01f232676d1e3c7f129071b2/git-cliff-core/src/remote/bitbucket.rs#L52-L58

**Comment by orhun** [comment]
Duplicate, see #891

**Comment by mcwarman** [comment]
I was looking at that one, it could be duplicate, but i think the scenarios are slightly different:

This one refers to commit not in the release effecting the users `is_first_time` status. 

Where as I believe #891 is when a previous releases logs `is_first_time` status is effected.

**Comment by orhun** [comment]
Hmm I see, I guess this one is also about using different branches and `--current` flag. Do you think fixing #891 would affect this in some way as well?

> I guess the fix is to make the current branch context available in the Remote struct so it can be used in the API call.

This makes sense, however, we should look into other remotes (GitLab,GitHub,Gitea) to see if it is possible. Any ideas?

**Comment by mcwarman** [comment]
Looks possible according to the various docs:

I think they're all effected looking at the API documentation GitLab and GitHub is possible less obvious to users as it defaults to the default branch.

## GitHub

Looks to default to the "default branch", but wouldn't work for users if they'd expect it to be running in different context.

```
GET /repos/{owner}/{repo}/commits
```

> Query Parameters
> `sha` SHA or branch to start listing commits from. Default: the repository’s default branch (usually main).

https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#list-commits

## GitLab

Looks to default to the "default branch", but wouldn't work for users if they'd expect it to be running in different context.

```
GET /projects/:id/repository/commits
```

> Query Parameters
> `ref_name` The name of a repository branch, tag or revision range, or if not given the default branch

https://docs.gitlab.com/ee/api/commits.html#list-repository-commits

## Gitea

Not entirely clear what the default behaviour is here from the documentation, but its either going to work like GitHub/GitLab or Bitbucket

```
GET /repos/{owner}/{repo}/commits
```

> Query Parameters
> `sha` SHA or branch to start listing commits from (usually 'master')

https://docs.gitea.com/api/1.20/#tag/repository/operation/repoGetAllCommits

## Bitbucket

Returns all commits in the repo in topological order (newest commit first). All branches and tags are included (similar to git log --all).

```
GET /repositories/{workspace}/{repo_slug}/commits
```

https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-commits-get

Alternate endpoint might be better `GET /repositories/{workspace}/{repo_slug}/commits/{revision}`

https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-commits-revision-get

**Comment by orhun** [comment]
Thanks for investigating this further! Then I guess we can add a `branch` field to the `remote` and update the API usages 👍🏼 

Would you be interested in working on this? My schedule is a bit tight these days (due to approaching EuroRust conference) so I might not get to it until next week.

**Comment by mcwarman** [comment]
Yeah I'd be interested, I'll try have a look, I created the scenario in the test repo here:

- Commit in other repo effect is_first_status: https://bitbucket.org/mcwarman-playground/git-cliff-readme-example/commits/2b61aa56962e50f64b507589448e04f9bc963c47
- Commit in default branch: https://bitbucket.org/mcwarman-playground/git-cliff-readme-example/commits/655ded3d6587d4d08ddb15054d50302c2ff61a81

**Comment by orhun** [comment]
Awesome

**Comment by mcwarman** [comment]
Hello, sorry for the delay and sorry for never getting around to looking at it, it quickly moved out of my rust ability.

I would have actually expected the `test-bitbucket-integration` test to fail (need fixing) if it was fixed as I'd added the scenario in there.

**Comment by mcwarman** [comment]
https://github.com/orhun/git-cliff/blame/9e4bd077b5a18922021afd8ffd671b7d7958ee5c/.github/fixtures/test-bitbucket-integration/expected.md#L8

`mcwarman` should be reported as first time contributor for this release.

**Comment by orhun** [comment]
That's strange... should we re-open this issue and take a look at it again? I don't remember the root cause of it and need to read it again...

**Comment by mcwarman** [comment]
It was all to do with the fact that the API that's being used to collect the commits returns all the commits of the repo rather than the commits for the branch itself.

```
git log --graph --all --decorate
```

```
* commit 655ded3d6587d4d08ddb15054d50302c2ff61a81 (HEAD -> master, origin/master, origin/HEAD)
| Author: mcwarman <7236500+mcwarman@users.noreply.github.com>
| Date:   Sun Oct 6 22:12:19 2024 +0100
|
|     feat(ci): add pipeline config
|
| * commit 2b61aa56962e50f64b507589448e04f9bc963c47 (origin/docs/example-typo)
|/  Author: mcwarman <7236500+mcwarman@users.noreply.github.com>
|   Date:   Sun Oct 6 22:11:56 2024 +0100
|
|       docs(example): fix typo
|
* commit df6aef41292f3ffe5887754232e6ea7831c50ba5
| Author: orhun <orhun@archlinux.org>
| Date:   Sun Jul 18 15:14:39 2021 +0300
|
|     feat(cache): use cache while fetching pages
|
...
```

**Comment by mcwarman** [comment]
In the e2e test, the following API is being used for bitbucket:

```
GET https://api.bitbucket.org/2.0/repositories/mcwarman-playground/git-cliff-readme-example/commits
```

But should be using:

```
GET https://api.bitbucket.org/2.0/repositories/mcwarman-playground/git-cliff-readme-example/commits/master
```

**Comment by mcwarman** [comment]
The challenge I had implementing it was plumbing the current branch through.

**Comment by orhun** [comment]
Oh, the branch specific remote fetching is now implemented in #1086 (it's not released yet).

So should we expect this issue to be fixed now? Can you test on `main`?
